fix(route53): vpc region in template overridden by stack region

daschaa opened the pull request:
Fixes #20496 

This PR implements the proposed change in #20496 - When a region is set in the vpc it is used in the CloudFormation template. Otherwise the region from the respective stack is used.

----

### All Submissions:

* [x] Have you followed the guidelines in our [Contributing guide?](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md)

### Adding new Unconventional Dependencies:

* [ ] This PR adds new unconventional dependencies following the process described [here](https://github.com/aws/aws-cdk/blob/master/CONTRIBUTING.md/#adding-new-unconventional-dependencies)

### New Features

* [x] Have you added the new feature to an [integration test](https://github.com/aws/aws-cdk/blob/master/INTEGRATION_TESTS.md)?
	* [x] Did you use `yarn integ` to deploy the infrastructure and generate the snapshot (i.e. `yarn integ` without `--dry-run`)?

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildProject89A8053A-LhjRyN9kxr8o
* Commit ID: ac4f9e267222960ce943f971f5d281d68b43ed6e
* Result: SUCCEEDED
* [Build Logs](https://xuogv3vm9j.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0f1d23cf-c947-4073-b0ef-5e2331f8d56b%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

daschaa commented:
@peterwoodworth Since you commented on the original issue - could you review this PR please? I would appreciate the feedback! ❤️

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 532ad404e40f8bd463d4ea632300410fec8c8d17
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=2e710c30-9326-4ad3-9d0e-480f96f84814%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

peterwoodworth commented:
Sure thing @daschaa, I'll try to get to this today 🙂

peterwoodworth reviewed (dismissed):
Thanks for the contribution @daschaa!

This is going to be a slightly larger PR than I had anticipated. We need to add some additional functionality elsewhere for this to work.

First, let's take a look at how our S3 module handles this:
[`Bucket.fromBucketAttributes()`](https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_aws-s3.Bucket.html#static-fromwbrbucketwbrattributesscope-id-attrs) has a [`region`](https://docs.aws.amazon.com/cdk/api/v1/docs/@aws-cdk_aws-s3.BucketAttributes.html#region) prop the user can pass in for when the region of the stack and bucket differ. This ends up being represented to the user through `IBucket.env.region`
https://github.com/aws/aws-cdk/blob/ab288785279bf3d70ad3329e27ad1f6d2397942e/packages/%40aws-cdk/aws-s3/lib/bucket.ts#L1588-L1591

`Vpc.fromVpcAttributes()` and `Vpc.fromLookup()` should both support this. However, neither of these methods currently support this. The former doesn't allow you to set the region at all, and the latter does, but it isn't set [when the construct is created](https://github.com/aws/aws-cdk/blob/ab288785279bf3d70ad3329e27ad1f6d2397942e/packages/%40aws-cdk/aws-ec2/lib/vpc.ts#L1176). Both of these functionalities will need to be implemented 🙂 

As for tests, let's create tests in the EC2 library to verify that the region is properly set with these methods. Additionally, for the hostedzone test, could we use one of these import methods instead of a mock construct?

peterwoodworth commented:
No worries @daschaa 

> it doesn't really test the functionality - just if it is passed correctly

Well, that is all we need to test here! Right? All we want to know is that the Vpc you create with `fromLookup` sets its `.env.region` prop to the one you pass in. Check out [this test file](https://github.com/aws/aws-cdk/blob/main/packages/@aws-cdk/aws-ec2/test/vpc.from-lookup.test.ts) - this is the file where you'll want to add the test and it has plenty of examples 🙂 

Let me know if you need any more help!

daschaa commented:
@peterwoodworth I can now continue working on this feature again. Sorry for the delay!
I have added the region prop the the `fromLookup` and `fromVpcAttributes` and used `fromVpcAttributes` in the test for the hosted zone. 
However I'm not really happy about the typing of the `LookedUpVpc`. I extended the `props`-property with region and decided to not introduce an own type. I thought it would be overkill to just implement it for the region prop. Is it ok like that? I would be happy to get feedback from your side and maybe some better ideas about the typing. Thanks in advance!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: fcc388d9a0bc02e90dbafca92b086e3fea61ac2b
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=9b01f743-f6e0-4fda-935b-88e1a275ef73%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: e8a10112353839d3a7671e2a1ff368c02d549ddb
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=8c48f8b2-5f5a-4f15-88e2-fcf59d6cf35c%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 66a3c4f215e06ca9f6fa126bf579da3662b7abbc
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=c6b36580-b613-4a65-8f39-30c44aa2f065%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

daschaa commented:
@peterwoodworth I just saw that I did not rebase this branch very well, but I think everything is fine. At least the changes still look good :) So please excuse me for this huge list of commits in this PR 😄

TheRealAmazonKendra commented:
Rebasing gets weird sometimes ¯\\\_(ツ)\_/¯

daschaa commented:
@mergifyio update

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 1a96ebc76c23c4b345005679d084b02cf35d4e62
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=48e85499-7fd2-44a3-a284-9acb5c78bf14%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

daschaa commented:
@TheRealAmazonKendra Is this mergify command new? Never saw this before and wondered if we should use it instead of the GitHub Update Branch command :)

TheRealAmazonKendra commented:
> @TheRealAmazonKendra Is this mergify command new? Never saw this before and wondered if we should use it instead of the GitHub Update Branch command :)

I use it because it doesn't dismiss previous reviews and will rerun workflows. I think I'm the only one who really uses the Mergify commands, though. Either way is fine.

TheRealAmazonKendra commented on packages/@aws-cdk/aws-ec2/lib/vpc.ts:
```diff
@@ -1983,7 +1995,9 @@ class ImportedVpc extends VpcBase {
   constructor(scope: Construct, id: string, props: VpcAttributes, isIncomplete: boolean) {
-    super(scope, id);
+    super(scope, id, {
+      region: props.region ?? undefined,
```
This is redundant. `region: props.region` means the same thing as `region: props.region ?? undefined` when `props.region` is potentially undefined.

```suggestion
      region: props.region,
```

TheRealAmazonKendra commented on packages/@aws-cdk/aws-ec2/lib/vpc.ts:
```diff
@@ -2032,8 +2046,10 @@ class LookedUpVpc extends VpcBase {
-    super(scope, id);
+  constructor(scope: Construct, id: string, props: cxapi.VpcContextResponse & { region?: string }, isIncomplete: boolean) {
+    super(scope, id, {
+      region: props.region ?? undefined,
```
Same comment as above. 
```suggestion
      region: props.region,
```

TheRealAmazonKendra reviewed (dismissed):
I've added some very preliminary feedback here. I'd also like to see an integration test for this change, please.

Oh, also, please take a look and see if you think there's a place where README documentation will need to be updated with this change. I'll trust your judgement on that.

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 3630c38267dad325306f1b3625bb451e709062bb
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=5261dfd5-2347-4abf-a691-5d5cb45e10f1%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TheRealAmazonKendra reviewed (dismissed):
Actually, I'm going to revise my earlier statement about using `@Mergifyio update`. Please do use it instead of merging via other methods. I'm finding it much easier to track when changes have been made on the PRs where that command is being used. Since we're trying to provide faster responses to PRs in general, this helps us be more efficient. (Also just putting this back into changes requested so that I'll see when you've made a code update).

daschaa commented:
@TheRealAmazonKendra I have added your suggested changes. Thanks for the feedback!
Regarding the integration tests: Do you mean Integration Tests for the `fromLookup` and `fromVpcAttributes` methods? I wonder if this would be easily possible since the snapshot would not do a real lookup, right? 🤔 
Or do you mean for the `addVpc` method?

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 8dfd079e9a1db33d52444f4ec558be6861fd938c
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=b6466983-fe3e-45bf-9e0b-91451e1f4919%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TheRealAmazonKendra commented:
New/updated integration tests actually create the resources and deploy them. They just don't run in a public account so you never see it happen. We also will see from the synthesized template that the region assigned to the VPC will be different from the one you assign for the stack. If you create the test using the `IntegTest` construct and then run `yarn integ --update-on-failed <filename>.js` then it will setup the resources, tear them back down, and produce the expected snapshot files to compare against later. Check out our new documentation on [integration tests](https://github.com/aws/aws-cdk/blob/main/INTEGRATION_TESTS.md) for more info.

TheRealAmazonKendra reviewed (changes requested):
I thought I commented this earlier but I can't find it anywhere. I tend to think that region should be added to `VpcContextResponse` and added to `attributes` more organically that way, instead of through the additional prop.

TheRealAmazonKendra commented:
> @TheRealAmazonKendra I have added your suggested changes. Thanks for the feedback! Regarding the integration tests: Do you mean Integration Tests for the `fromLookup` and `fromVpcAttributes` methods? I wonder if this would be easily possible since the snapshot would not do a real lookup, right? 🤔 Or do you mean for the `addVpc` method?

I just realized that I only kind of answered this question. I'd want to see integ tests on `fromLookup` and `fromVpcAttributes`. How I see that working is to create two stacks in your test. One in region-a that creates the VPC and one in region-b that uses `fromLookup` and `fromVpcAttributes` to import it. Only list the stack that uses these functions in your `testCases` array.

daschaa commented:
@TheRealAmazonKendra Sorry for not working on this consistently - I have a lot on my plate right now.
I have now added the integration tests for `fromLookup` and `fromVpcAttributes` as you suggested and I moved the region prop to the `VpcContextResponse`. 
I look forward to your feedback :)

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 49b23d7aa00752eccd8891ce1ca78ce45aab68c2
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=23130e6c-8c67-41d1-b55d-3df6992ed3ae%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 9e0c5336fd7022d660fafc43d792cd703f2da23a
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=95b7cb83-64b4-4815-ba9a-5f61e2d3a7e0%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TheRealAmazonKendra commented:
> @TheRealAmazonKendra Sorry for not working on this consistently - I have a lot on my plate right now. I have now added the integration tests for `fromLookup` and `fromVpcAttributes` as you suggested and I moved the region prop to the `VpcContextResponse`. I look forward to your feedback :)

No need to be sorry at all, you're contributing your time and resources to this and we're super grateful for it!

TheRealAmazonKendra commented:
FYI - I haven't forgot this PR but I do think there's more refactoring that needs to be done so I'm taking a deeper dive into the code to make sure my suggestions aren't off base.

daschaa commented:
@TheRealAmazonKendra Any updates on this one? :)

TheRealAmazonKendra commented:
@Mergifyio update

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: b515e6a811eda0bd3d250250e62dc9e4d1862f8c
* Result: FAILED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=440aa0c0-2e21-420f-8be5-b5ae7ab56eab%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

daschaa commented:
@Mergifyio update

daschaa commented on packages/@aws-cdk/aws-ec2/lib/vpc.ts:
```diff
@@ -1203,7 +1210,12 @@ export class Vpc extends VpcBase {
-    return new LookedUpVpc(scope, id, attributes || DUMMY_VPC_PROPS, attributes === undefined);
+    const lookedUpVpcProps = {
+      ...attributes,
+      region: options.region,
```
I think you are right, this assignment is redundant. I have now removed it and the integration tests still work as expected. Thanks for pointing this out :)

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 9063de03a468c5879e3959308dbdfcb0345db82a
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4767ecd8-99fe-4411-aba6-45983476ad30%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 9d0ed5b231d042b28849ad84abe0788e098d8291
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=0afb72bc-2ac8-48a3-b5c1-a087575d0a7e%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

TheRealAmazonKendra reviewed (approved):
Thanks for all your work on this and patience in the long delay between feedback. This looks great!

aws-cdk-automation commented:
### AWS CodeBuild CI Report

* CodeBuild project: AutoBuildv2Project1C6BFA3F-wQm2hXv2jqQv
* Commit ID: 84aca59854ddae062e21f082604eecdedd4dfd9e
* Result: SUCCEEDED
* [Build Logs](https://2cl995f30a.execute-api.us-east-1.amazonaws.com/Prod/buildlogs?key=4754ea0f-b05b-4571-8208-a1ac153d64bf%2Fbuild.log) (available for 30 days)

*Powered by [github-codebuild-logs](https://serverlessrepo.aws.amazon.com/applications/arn:aws:serverlessrepo:us-east-1:277187709615:applications~github-codebuild-logs), available on the [AWS Serverless Application Repository](https://aws.amazon.com/serverless/serverlessrepo/)*


<!--
CREATED BY GITHUB-CODEBUILD-LOGS
-->

michaelciccarelli commented:
Can anyone follow up on what I'm seeing? I'm running 2.44.0 and this doesn't seem like it's working yet:

       if env.region == "us-east-1" and not self.bcp_env:
            useast2vpcid = os.getenv("VPC_ID_US_EAST_2", "MISSING__VPC_ID_US_EAST_2_FOR_HIGHAVAILABILITY")
            useast2vpc = Vpc.from_lookup(self, "vpcLookup-useast2", vpc_id=useast2vpcid, region="us-east-2")
            my_hosted_zone = route53.HostedZone(self, APP_NAME + "-route53",
                zone_name = subdomain + ".essent.us",
                vpcs = [ self.vpc,  useast2vpc  ]
            )

When I run it, the synth is fine but apply shows that it CDKformation doesn't have permission to add this vpc.. the issue is the vpc is showing up as in us-east-1 even though we're specifying us-east-2.. The following code fixes the issue with calling Cfn directly.... 

       if env.region == "us-east-1" and not self.bcp_env:
            useast2vpcid = os.getenv("VPC_ID_US_EAST_2", "MISSING__VPC_ID_US_EAST_2_FOR_HIGHAVAILABILITY")
#            useast2vpc = Vpc.from_lookup(self, "vpcLookup-useast2", vpc_id=useast2vpcid, region="us-east-2")
#            my_hosted_zone = route53.HostedZone(self, APP_NAME + "-route53",
            my_hosted_zone = route53.CfnHostedZone(self, APP_NAME + "-route53",
                name = subdomain + ".essent.us",
#                vpcs = [ self.vpc ]
                vpcs = [
                    route53.CfnHostedZone.VPCProperty(vpc_id=self.vpc_id, vpc_region = env.region),
                    route53.CfnHostedZone.VPCProperty(vpc_id=useast2vpcid, vpc_region = "us-east-2")
                ]
            )
